samples/nrf9160/http_application_update/src/main.c

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <stdint.h>
+#include <zephyr.h>
+#include <flash.h>
+#include <bsd.h>
+#include <nrf_socket.h>
+#include <download_client.h>
+#include <logging/log.h>
+#include <gpio.h>
+#include <flash_map.h>
+#include <dfu/mcuboot.h>
+
+LOG_MODULE_REGISTER(main);
+
+#define LED_PORT	LED0_GPIO_CONTROLLER
+
+static struct	device	*flash_dev;
+static K_THREAD_STACK_DEFINE(dfu_thread_stack, 1024);
+static bool is_flash_page_erased[FLASH_PAGE_MAX_CNT];
+
+static struct	k_thread dfu_thread;
+static bool		start_dfu;
+static struct	gpio_callback gpio_cb;
+
+static int app_download_client_event_handler(struct download_client *const dfu,
+			enum download_client_evt event, u32_t status);
+
+
+static struct download_client dfu = {
+	.host = CONFIG_DOWNLOAD_HOST,
+	.resource = CONFIG_DOWNLOAD_FILE,
+	.callback = app_download_client_event_handler
+};
+
+
+/**@brief Recoverable BSD library error. */
+void bsd_recoverable_error_handler(uint32_t err)
```

Comment:
If those error functions are not overwritten, the default function will hard fault.
You could maybe add a print or assert instead fo a busy loop, as done in other examples, for instance https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/master/samples/nrf9160/at_client/src/main.c#L13.

Reply:
There was a print but it was removed by request of @gbakke. Printing will not work inside these functions.

Reply:
@osaether @gbakke can you please elaborate here? Are you saying `printk` is not working?

Reply:
`printk()` works with default backend (as used in asset tracker and gateway), and I thought it should work here as well with both `CONFIG_LOG_PRINTK` and `CONFIG_LOG_IMMEDIATE` enabled, but I could definitely be wrong. At least it wouldn't be output with only `CONFIG_LOG_PRINTK` enabled before exiting the error handler.

Reply:
I will add printk to bsd_recoverable_error_handler but leave bsd_irrecoverable_error_handler as is. Accordning to @gbakke bsd_irrecoverable_error_handler  can happen during device init, before the UART device is initialized.

Reply:
I don't think that is the case: `bsdlib` is initialized in `bsd_os.c` with `POST_KERNEL` priority, the UARTE driver is initialized with `PRE_KERNEL` priority. We should definitely add a `printk` in the non-recoverable error handler.

Reply:
@lemrey printk in both now